update docker research (#67)
thoth-tech/documentation: docs/OnTrack/Jupyter Notebook/Docker-Documentation-Research-T1-2022.md

# Docker Documentation and Research (WIP) T1-2022

[Back to Jupyter Notebook Documentation Index](Index.md)

## Intro

> The Jupyter Notebook conversion feature will happen via processes that occur inside Docker

Reviewer: Reword:

```suggestion
The Jupyter Notebook conversion feature will occur via processes inside Docker
```

containers. In this document we will discuss the main architecture of this structure in relation to
OnTrack.

> OnTrack is deployed with two main containers that will always be running:

Reviewer: Reword:

```suggestion
OnTrack is deployed through two main containers, they are:
```

- The front-end container which hosts the logic behind the user interface of OnTrack (mostly
  unrelated to this feature)
- The back-end container which hosts the logic related to database interactions and other processes

> To achieve the Jupyter Notebook conversion feature, we will be using a new container that contains

Reviewer: Reword:

```suggestion
A new container will be created to achieve the Jupyter Notebook conversion feature. This contains
```

conversion, and performing the conversion process within this container. This allows us to have a
standalone software package that is extendable if required, for example if new Python libraries are
required by a unit. It also means that we are able to create as many standalone containers as
required for different conversion processes, such as:

- Docx to PDF conversion using Apache POI
- Powerpoint presentation to PDF conversion using Apache POI

## Overview of Conversion Process

[<img src="docker_flow.png" />](docker_flow.png)

When the OnTrack front-end sends a new file to the OnTrack back-end, the back-end will be able to
determine the file type, and if the file needs to be converted. If the OnTrack back-end recieves a
.ipynb file, it will run the Jupyter to PDF conversion container to perform the conversion process.

This is done via a shell command that does several things:

1. Firstly, it ensures that the file to be converted is renamed to "input.ipynb".
2. It then instructs the container to run AND it mounts the file's directory as a Docker volume. As
   the container runs it will execute the conversion process via its `ENTRYPOINT` command. The

>    converted file will be outputted to the volume. After the container has run, it will be removed.

Reviewer: Typo

```suggestion
   converted file will be output to the volume. After the container has run, it will be removed.
```

3. Finally, it removes the temporary files to ensure there is always free space.

> The `docker run` command allows us several options for this type of use-case: we are able to specify

Reviewer: ```suggestion
The `docker run` command provides several options for this type of use-case: we are able to specify
```

a volume using the `-v` option, and we are able to ask that the container is removed after it has
finished its process with the `--rm` option.

**Note: we may need to have some process in place to read whether the conversion was a success or a
failure.**

## Conversion Container Requirements

> For a container to work in this architecture, there are some requirements that have to be met:

Reviewer: Reword:

```suggestion
For a container to function in this architecture, there are requirements to be met:
```

- The container will run the conversion as its `ENTRYPOINT` command _without any user input
  required_.
- The container will always look in a single directory (mounted as a volume when the container is
  run) for **one** file that will be called "input" (+ whatever file extension is required).
- The container will always output the converted file to the **same** directory (mounted as a volume
  when the container is run) and call it "output.pdf".

> These requirements allow for the containers themselves to remain isolated, while the logic about
> which files need to be converted can be handled seperately by OnTrack's backend. This is necessary

Reviewer: Reword:

```suggestion
These requirements allow the containers to remain isolated while the file conversion logic is handled seperately by the OnTrack backend. This is necessary
```

# Docker Documentation and Research (WIP) T1-2022

[Back to Jupyter Notebook Documentation Index](Index.md)

## Intro

As we know, the Jupyter Notebook conversion feature will happen via processes that occur inside

> Docker containers. In this document we will discuss the main architecture of this structure in

Reviewer 2: I would remove the "as we know". Assume this documentation is being read by people who have never worked on this project before and they don't know anything

relation to OnTrack.

OnTrack is deployed with two main containers that will always be running:

- the frontend container which hosts the logic behind the user interface of OnTrack (mostly
  unrelated to this feature)

> - the backend container which hosts the logic related to database interactions and other processes

Reviewer 3: To me, it might be better to have frontend and backend capitalized and maybe have them as 'Front-end' and 'Back-end' with a '-' in the middle, just to ensure readability. Just a suggestion though.

Author: Thanks @mfclark1 , I've made this change, although have kept 'back-end' anmd 'front-end' lowercase.

> To achieve the Jupyter conversion feature, we will be using a new container that contains all of the

Reviewer 3: Maybe "the Jupyter **Notebook** conversion feature...", just so the reader understands that is what you are talking about. Whilst it may not be the case of every reader, some readers might see 'Jupyter Notebook' and 'Jupyter' and think they are two separate things. Just a suggestion though.

> dependencies (Python, TeX, etc.) needed for a .ipynb to PDF conversion, and performing the

Reviewer 3: Might need to explain what a ".ipynb" is. Some might understand that it is the file type for the Jupyter Notebooks but other's may not. Just a suggestion though.

conversion process within this container. This allows us to have a standalone software package that
is extendable if required, for example if new Python libraries are required by a unit. It also means
that we are able to create as many standalone containers as required for different conversion
processes, such as:

- docx to PDF conversion using Apache POI
- powerpoint presentation to PDF conversion using Apache POI

> This will allow us to add a wide range of compatability in the future, regardless of the software

Reviewer 3: Not sure of the use of the word "compatability", maybe "conversion capabilities"? I'm not too sure myself. 
Just a suggestion though

requirements.

## Overview of Conversion Process

[<img src="docker_flow.png" />](docker_flow.png)

> When the OnTrack frontend sends a new file to the OnTrack backend, the backend will be able to

Reviewer 2: a little picky - but making sure your sentences start with a capital and frontend/backend is front-end/back-end

Author: Picky is good! Thanks @ginge270 changes have been made.

> determine the file type and if it needs to be converted. If the OnTrack backed recieves a .ipynb

Reviewer 3: Not sure about "... and if it needs to be converted. ..." as it doesn't really say what might need to be converted. I was thinking maybe something along the lines of "... determine the file type, and if the file needs to be converted. ...". I'm not too sure whether I'm being too picky. Just a suggestion though.

> determine the file type and if it needs to be converted. If the OnTrack backed recieves a .ipynb

Reviewer 3: Misspelling: "If the OnTrack **backed"**..."

> file, it will 'spin up' the Jupyter to PDF conversion container to perform the conversion process.

Reviewer 3: Reader may not understand what you mean by "spin up". Explanation of this may improve readability. Just a suggestion though.

a volume using the `-v` option, and we are able to ask that the container is removed after it has
finished its process with the `--rm` option.

**Note: we may need to have some process in place to read whether the conversion was a success or a
failure.**

## Conversion Container Requirements

For a container to work in this architecture, there are some requirements that have to be met:

- The container will run the conversion as its `ENTRYPOINT` command _without any user input
  required_.
- The container will always look in a single directory (mounted as a volume when the container is
  run) for **one** file that will be called "input" (+ whatever file extension is required).
- The container will always output the converted file to the **same** directory (mounted as a volume
  when the container is run) and call it "output.pdf".

These requirements allow for the containers themselves to remain isolated, while the logic about
which files need to be converted can be handled seperately by OnTrack's backend. This is necessary
to follow the Single-responsibility Principle: the container itself is responsible for only one

> task - that is, performing the conversion process.

Reviewer 3: Comments for docker_flow.png:
- It may be worth changing 'frontend' and 'backend' to 'Front-end' and 'Back-end' respectively, to ensure readability and keep the same spelling throughout.
- Not sure what **POST** means on POST input file (arrow left of OnTrack Backend box). Not sure if this needs to be explained or not. 
- Whilst I understand what "return success or failure" means in that it says whether the conversion was a success or not, other viewers may not understand. It might be worth saying something like "return success or failure of conversion process", just so readers know what it indicates.

Just suggestions though.
